src/Microsoft.ML.Data/StaticPipe/StaticPipeUtils.cs

```diff
@@ -366,5 +367,97 @@ public BidirectionalDictionary()
+        /// instance <see cref="Indices"/>, returns one of the <see cref="PipelineColumn"/> instances stored
+        /// therein, which the component can use to do specific operations.
+        /// </summary>
+        /// <typeparam name="T">The shape type.</typeparam>
+        public sealed class IndexHelper<T>
+        {
+            /// <summary>
+            /// An instance of the shape type whose items can be used to index to find the names of column.
+            /// </summary>
+            public T Indices { get; }
+
+            /// <summary>
+            /// Maps the items inside <see cref="Indices"/> to the names of the associated data's column's name.
+            /// Components can use this structure, but it may be preferable to use <see cref="Get(PipelineColumn, IExceptionContext)" />
+            /// to ensure uniformity in the exception messages, if possible.
+            /// </summary>
+            private ImmutableDictionary<PipelineColumn, string> Map { get; }
+
+            /// <summary>
+            /// Performs a lookup on <see cref="Map"/>. If the key is not present this will throw an exception
+            /// more generally helpful in context than that of a direct failure of index on <see cref="Map"/>.
+            /// </summary>
+            /// <param name="key">The column to look up.</param>
+            /// <param name="ectx">The optional exception context.</param>
+            /// <returns>If successful the name of the column.</returns>
+            public string Get(PipelineColumn key, IExceptionContext ectx = null)
+            {
+                Contracts.CheckValueOrNull(ectx);
+                ectx.CheckValue(key, nameof(key));
+                if (!Map.TryGetValue(key, out string name))
+                {
+                    // The most obvious reason this might happen is if the user did something like try to attempt to
+                    // apply an estimator inside the index delegate, which obviously will not work.
+                    throw ectx.ExceptParam(nameof(key), "Column does not appear to be valid for this structure. " +
+                        "Please use columns in the provided indexing object without attempting modification.");
+                }
+                return name;
+            }
+
+            /// <summary>
+            /// Constructor for the index helper. Note that any public or component code will instead use
+            /// <see cref="GetIndexer{T}(SchemaBearing{T})"/>, to fetch the lazily constructed instance of this
+            /// object, since its construction is somewhat expensive.
+            /// </summary>
+            /// <param name="schematized">Constructs the helper for this object.</param>
+            internal IndexHelper(SchemaBearing<T> schematized)
+            {
+                Indices = StaticPipeInternalUtils.MakeAnalysisInstance<T>(out var rec);
+                // We don't really care about this, but we need to work over something.
```

Comment:
>We don't really care about this, but we need to work over something. [](start = 19, length = 68)

I'm sorry, I completely don't understand what is going on. The comment adds negative clarity.

Reply:
I see, you're constructing a method just to manufacture a `ParameterInfo`. Okay. Then you're calling `GetNamesValues`, but you only ever need values.

At least add a comment to `GetNamesValues` please

---
In reply to: [216415488](https://github.com/dotnet/machinelearning/pull/869#discussion_r216415488) [](ancestors = 216415488)

Reply:
Very well.

---
In reply to: [216416576](https://github.com/dotnet/machinelearning/pull/869#discussion_r216416576) [](ancestors = 216416576,216415488)